{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-29T13:27:00.076127Z",
     "start_time": "2019-04-29T13:26:57.852409Z"
    }
   },
   "outputs": [],
   "source": [
    "import SinglePython\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-29T13:27:07.799312Z",
     "start_time": "2019-04-29T13:27:00.084080Z"
    }
   },
   "outputs": [],
   "source": [
    "# Uploading an existing reference dataset. \n",
    "# As an option, a reference dataset generated by the user can also be used. \n",
    "refDataset = pd.read_csv(\"./Data/Reference/hpca_data.csv\", sep=\";\", decimal = \",\", index_col = 0)\n",
    "refDataset = refDataset.astype('float')\n",
    "annot = pd.read_csv(\"./Data/Reference/hpca_annot.csv\",sep=\";\", index_col = 0)\n",
    "del annot[\"level1\"]\n",
    "annot.columns = [\"level1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-29T13:27:13.388519Z",
     "start_time": "2019-04-29T13:27:07.858561Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# An example for a single cell experiment\n",
    "# The input is the file path which the count data is present. \n",
    "# 10X datasets are automatically identified. \n",
    "# For the count data that is present in a csv file. Please make sure the file has \"expression\" keyword in file name in order\n",
    "# for the program to identify it.\n",
    "data =utils.readSCData(\"./Data/Datasets/filtered_feature_bc_matrix/\",200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-29T13:27:13.514152Z",
     "start_time": "2019-04-29T13:27:13.440343Z"
    }
   },
   "outputs": [],
   "source": [
    "# A singlePython object is created. \n",
    "# This object keeps the original single cell data as an AnnData format (scanpy format).\n",
    "obj = SinglePython.SinglePythonObject(refDataset=refDataset,refAnnot=annot,scData=data,fine_tuning=True,tuning_by=\"threshold\",multiProcess=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-04-29T13:27:00.431Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\murat_gga8ya6\\Desktop\\Thesis\\SingleRToPython\\tuningMulti.py:43: FutureWarning: by argument to sort_index is deprecated, please use .sort_values(by=...)\n",
      "  [d.update({i:np.sort(scores.sort_index(by=i,ascending=False).index.values[0:n])}) for i in scores.columns]\n"
     ]
    }
   ],
   "source": [
    "# The function below starts the annotation procedure.\n",
    "# The result is added as a observation to the original single cell AnnData object.\n",
    "obj.annotateCellTypes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-04-29T13:27:04.126Z"
    }
   },
   "outputs": [],
   "source": [
    "# The result can be obtained from AnnData object for further examinations.\n",
    "obj.scData.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Below here is for further examinations. ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You can observe the gene expression distribution over the samples to determine cells with minimum and maximum number of genes.\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "def histMatrix(matrix):\n",
    "    list_of_lists = matrix.values.tolist()\n",
    "    list_of_fluxes = [item for sublist in list_of_lists for item in sublist]\n",
    "    sns.set(style=\"white\", palette=\"muted\",color_codes=True)\n",
    "    plot = sns.distplot(list_of_fluxes)\n",
    "histMatrix(pd.DataFrame(obj.sc_data.obs.n_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-05T17:26:40.685537Z",
     "start_time": "2019-04-05T17:26:40.409254Z"
    }
   },
   "outputs": [],
   "source": [
    "# The functions for preprocessing of single cell data matrix. You can skip to the next notebook cell.\n",
    "class CellMetricsFilter:\n",
    "\n",
    "    def __init__(self, minUMICounts, maxUMICounts, minGenesPerCell, minGeneOccurenceInCell, maxPercentMito = .1, regressOut = []):\n",
    "\n",
    "        \n",
    "\n",
    "        self.minUMICounts = minUMICounts\n",
    "        self.maxUMICounts = maxUMICounts\n",
    "        self.minGenesPerCell = minGenesPerCell\n",
    "        self.minGeneOccurenceInCell = minGeneOccurenceInCell\n",
    "        self.maxPercentMito = maxPercentMito\n",
    "        self.regressOut = regressOut\n",
    "\n",
    "    def filter(self, data, sampleName):\n",
    "\n",
    "        matplotlib.rcParams['figure.figsize'] = [3, 3]\n",
    "\n",
    "        data.var_names_make_unique()\n",
    "\n",
    "        print(data)\n",
    "        mito_genes = [name for name in data.var_names if name.startswith(\"MT-\")]\n",
    "        data.obs['percent_mito'] = np.ravel(np.sum(data[:, mito_genes].X, axis=1)) / np.ravel(np.sum(data.X, axis=1))\n",
    "        data.obs['n_counts'] = np.ravel(np.sum(data.X, axis=1))\n",
    "\n",
    " \n",
    "\n",
    "        sc.pp.filter_cells(data, min_genes=0) # Hack to generate the n_genes column\n",
    "        sc.pp.filter_genes(data, min_cells=0)\n",
    "\n",
    " \n",
    "        priorFilteringCellCount = data.n_obs\n",
    "        priorFilteringGeneCount = len(data.var[data.var[\"n_cells\"] > 0])\n",
    "\n",
    "        print(\"Pre-filtering data characteristics for \" + sampleName)\n",
    "        sc.pl.violin(data, ['n_genes', 'n_counts', 'percent_mito'],\n",
    "\n",
    "                 jitter=0.4, multi_panel=True)\n",
    "        \n",
    "\n",
    "        print(\"Filtering genes occuring in <\" + str(self.minGeneOccurenceInCell) + \" cells.\")\n",
    "        sc.pp.filter_genes(data, min_cells=self.minGeneOccurenceInCell)\n",
    "\n",
    "        print(\"Filtering cells with <\" + str(self.minGenesPerCell) + \" genes.\")\n",
    "        sc.pp.filter_cells(data, min_genes=self.minGenesPerCell)\n",
    "\n",
    "        sc.pl.scatter(data, x='n_counts', y='percent_mito')\n",
    "        sc.pl.scatter(data, x='n_counts', y='n_genes')\n",
    "\n",
    "        print(\"Filtering cells based on >\" + str(self.minUMICounts) + \" and <\" + str(self.maxUMICounts) + \" UMI counts.\")\n",
    "        data = data[data.obs['n_counts'] > self.minUMICounts, :]\n",
    "        data = data[data.obs['n_counts'] < self.maxUMICounts, :]\n",
    "\n",
    "        print(\"Filtering cells based on mitochondrial ratio <\" + str(self.maxPercentMito) + \".\")\n",
    "        data = data[data.obs['percent_mito'] < self.maxPercentMito, :]\n",
    "\n",
    "        data.raw =  data.copy()\n",
    "        data = data[:,[i for i in data.var_names if not \"TotalSeq\" in i ]]\n",
    "\n",
    "        postFilteringCellCount = data.n_obs\n",
    "        postFilteringGeneCount = len(data.var[data.var[\"n_cells\"] > 0])\n",
    "\n",
    "        sc.pp.normalize_per_cell(data, counts_per_cell_after=1e4)\n",
    "        sc.pp.log1p(data)\n",
    "        sc.pp.highly_variable_genes(data, flavor = \"seurat\", inplace = True)\n",
    "\n",
    "        filter_result = sc.pp.highly_variable_genes(data, flavor = \"seurat\", inplace = False)\n",
    "\n",
    "        sc.pl.highly_variable_genes(filter_result)\n",
    "\n",
    "        try:\n",
    "\n",
    "            sc.tl.pca(data)\n",
    "            sc.pl.pca(data)\n",
    "\n",
    "        except ValueError:\n",
    "            print(\"Could not perform pca.\")\n",
    "\n",
    "        sc.pp.filter_genes(data, min_counts=1) # Remove unwanted zero count genes\n",
    "\n",
    "        if not self.regressOut: # but if we put nothing it gives error??\n",
    "            sc.pp.regress_out(data, self.regressOut)\n",
    "\n",
    "        sc.pp.scale(data, max_value=10)\n",
    "        \n",
    "\n",
    "        sc.pl.scatter(data, x='n_counts', y='percent_mito')\n",
    "        sc.pl.scatter(data, x='n_counts', y='n_genes')\n",
    "\n",
    "        try:\n",
    "            sc.tl.pca(data)\n",
    "            sc.pl.pca(data)\n",
    "\n",
    "        except ValueError:\n",
    "            print(\"Could not perform pca.\")\n",
    "\n",
    "        print(\"Post-filtering data characteristics for \" + sampleName)\n",
    "        sc.pl.violin(data, ['n_genes', 'n_counts', 'percent_mito'],\n",
    "            jitter=0.4, multi_panel=True, use_raw = False)\n",
    "\n",
    "        print(data)\n",
    "        return data\n",
    "    \n",
    "import os\n",
    "import matplotlib\n",
    "import scanpy.api as sc\n",
    "import scipy.io\n",
    "import numpy as np\n",
    "\n",
    "def readData(path):    \n",
    "    matplotlib.rcParams['figure.figsize'] = [3, 3]\n",
    "    \n",
    "    data = sc.read(path + 'matrix.mtx', cache=True).transpose()\n",
    "    data.var_names = np.genfromtxt(path + 'genes.tsv', dtype=str)[:, 1]\n",
    "    data.obs_names = np.genfromtxt(path + 'barcodes.tsv', dtype=str)\n",
    "    mito_genes = [name for name in data.var_names if name.startswith(\"MT-\")]\n",
    "    obj = CellMetricsFilter(500, 25000, 0, 3,mitoGenes=mito_genes,regressOut=True)\n",
    "    return obj.filter(data,\"Sample\")\n",
    "\n",
    "def cluster(data):\n",
    "    sc.tl.pca(data)\n",
    "    sc.pp.neighbors(data)\n",
    "    sc.logging.print_memory_usage()\n",
    "    sc.tl.umap(data)\n",
    "    sc.logging.print_memory_usage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-06T16:09:37.111572Z",
     "start_time": "2019-04-06T16:09:36.958965Z"
    }
   },
   "outputs": [],
   "source": [
    "# Application of single cell preprocessing.\n",
    "filter_obj = CellMetricsFilter(500, 2e8, 0, 3,regressOut=True)\n",
    "obj.sc_data = filter_obj.filter(obj.scData,\"Sample\")\n",
    "cluster(obj.scData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-06T16:14:21.994031Z",
     "start_time": "2019-04-06T16:14:21.989056Z"
    }
   },
   "outputs": [],
   "source": [
    "# Umap of cell classes.\n",
    "matplotlib.rcParams['figure.figsize'] = [10,10]\n",
    "sc.pl.umap(obj.scData, color = \"cell_type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-06T16:25:06.235859Z",
     "start_time": "2019-04-06T16:25:06.093239Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-06T11:50:50.134507Z",
     "start_time": "2019-04-06T11:49:51.412030Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-06T11:52:21.039997Z",
     "start_time": "2019-04-06T11:52:20.538784Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:58:48.187817Z",
     "start_time": "2019-02-10T16:58:48.176869Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:58:48.199813Z",
     "start_time": "2019-02-10T16:58:48.189809Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:58:55.604636Z",
     "start_time": "2019-02-10T16:58:48.201789Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:58:55.611644Z",
     "start_time": "2019-02-10T16:58:55.606631Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:59:12.888690Z",
     "start_time": "2019-02-10T16:58:55.613613Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:59:12.931575Z",
     "start_time": "2019-02-10T16:59:12.890658Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:59:12.938532Z",
     "start_time": "2019-02-10T16:59:12.933545Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:59:12.955518Z",
     "start_time": "2019-02-10T16:59:12.941524Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:59:13.070592Z",
     "start_time": "2019-02-10T16:59:12.957480Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:59:13.244128Z",
     "start_time": "2019-02-10T16:59:13.072562Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-10T16:59:13.872132Z",
     "start_time": "2019-02-10T16:59:13.246099Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-06T15:25:13.459555Z",
     "start_time": "2019-02-06T15:25:13.448432Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-02-05T11:07:03.728Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-29T12:11:13.036968Z",
     "start_time": "2019-04-29T12:11:13.025001Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-04-29T12:10:22.147459Z",
     "start_time": "2019-04-29T12:10:22.133486Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-02-05T11:07:03.755Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-02-05T11:07:03.757Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-02-05T16:07:21.700119Z",
     "start_time": "2019-02-05T16:07:21.693138Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-04-07T09:58:37.796Z"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    d={}\n",
    "    [d.update({i:np.sort(scores.sort_index(by=i,ascending=False).index.values[0:n])}) for i in scores.columns]\n",
    "    while(n>1):\n",
    "        unique_types=[list(x) for x in set(tuple(x) for x in d.values())]\n",
    "        for i in unique_types:\n",
    "            cols=[j for j in d.keys() if list(d.get(j))==i]\n",
    "            top_labels=i\n",
    "            res=_FineTuneRoundByN(sc_data,refDataset,annot,top_labels,de,cols)\n",
    "            [d.update({cols[t]:res[t]}) for t in range(len(cols))]\n",
    "        n=n-1\n",
    "    return pd.DataFrame(d,index=[\"final_annotations\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    d={}\n",
    "    [d.update({i:np.sort(scores.sort_index(by=i,ascending=False).index.values[0:n])}) for i in scores.columns]\n",
    "    while(n>1):\n",
    "        unique_types=[list(x) for x in set(tuple(x) for x in d.values())]\n",
    "        d=pd.DataFrame(d)\n",
    "        d_temp={}\n",
    "        for i in unique_types:\n",
    "            cols=d[d.isin(i)].T.dropna(how=\"any\").index.values\n",
    "            top_labels=i\n",
    "            res=_FineTuneRoundByN(sc_data,refDataset,annot,top_labels,de,cols)\n",
    "            [d_temp.update({cols[t]:res[t]}) for t in range(len(cols))]\n",
    "        n=n-1\n",
    "        d=d_temp\n",
    "    return pd.DataFrame(d,index=[\"final_annotations\"])"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
